File: PERSISTENCIA DE OBJETOS/Ejercicio 1/codigo/codigo.java
//5.Crea una clase Estudiante con nombre, nota_1, nota_2
//  a) Agrega un método para calcular el promedio
//  b) Agrega un método para verificar si aprobó (promedio >=6)
//  c) Crea tres estudiantes, muestra sus promedios y si aprobaron
// Empleado.java
import java.io.Serializable;

public class Empleado implements Serializable {
    private String nombre;
    private float salario;

    public Empleado(String nombre, float salario) {
        this.nombre = nombre;
        this.salario = salario;
    }

    // Getters
    public String getNombre() { return nombre; }
    public float getSalario() { return salario; }

    @Override
    public String toString() {
        return "Empleado{" + "nombre='" + nombre + '\'' + ", salario=" + salario + '}';
    }
}

import java.io.*;

public class ArchivoEmpleado {
    private String nombreArchivo;

    public ArchivoEmpleado(String nombreArchivo) {
        this.nombreArchivo = nombreArchivo;
    }

    public void guardarEmpleado(Empleado e) {
        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(nombreArchivo, true))) {
            oos.writeObject(e);
            System.out.println("Empleado guardado: " + e.getNombre());
        } catch (IOException ex) {
            System.err.println("Error al guardar empleado: " + ex.getMessage());
        }
    }

    public Empleado buscaEmpleado(String nombre) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nombreArchivo))) {
            while (true) {
                Empleado emp = (Empleado) ois.readObject();
                if (emp.getNombre().equalsIgnoreCase(nombre)) {
                    return emp;
                }
            }
        } catch (EOFException e) {
            System.out.println("Búsqueda completada");
        } catch (IOException | ClassNotFoundException ex) {
            System.err.println("Error al buscar empleado: " + ex.getMessage());
        }
        return null;
    }

    public Empleado mayorSalario(float sueldo) {
        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(nombreArchivo))) {
            while (true) {
                Empleado emp = (Empleado) ois.readObject();
                if (emp.getSalario() > sueldo) {
                    return emp;
                }
            }
        } catch (EOFException e) {
            System.out.println("Búsqueda completada");
        } catch (IOException | ClassNotFoundException ex) {
            System.err.println("Error al buscar empleado: " + ex.getMessage());
        }
        return null;
    }

    public static void main(String[] args) {
        ArchivoEmpleado archivo = new ArchivoEmpleado("empleados.dat");
        archivo.guardarEmpleado(new Empleado("Juan Pérez", 2500.50f));
        archivo.guardarEmpleado(new Empleado("María Gómez", 3200.75f));
        archivo.guardarEmpleado(new Empleado("Carlos Ruiz", 1800.25f));
        Empleado encontrado = archivo.buscaEmpleado("maría gómez");
        System.out.println("Empleado encontrado: " + encontrado);
        Empleado mayorSalario = archivo.mayorSalario(2000.0f);
        System.out.println("Empleado con salario > 2000: " + mayorSalario);
    }
}

File: PERSISTENCIA DE OBJETOS/Ejercicio 1/codigo/codigo.py
#1. Sea el siguiente diagrama de clases:
#  a) Implementa el método guardarEmpleado(Empleado e) para almacenarempleados.
#  b) Implementa buscaEmpleado(String n) a traves del nombre, para ver los datosdel Empleado n.
#  c) Implementa mayorSalario(float sueldo), que devuelva al primer empleado consueldo mayor al ingresado.
import pickle

class Empleado:
    def __init__(self, nombre, salario):
        self.nombre = nombre
        self.salario = salario
    
    def __str__(self):
        return f"Empleado(nombre='{self.nombre}', salario={self.salario})"

class ArchivoEmpleado:
    def __init__(self, nombre_archivo):
        self.nombre_archivo = nombre_archivo
    
    def guardar_empleado(self, empleado):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                empleados = pickle.load(f)
        except (FileNotFoundError, EOFError):
            empleados = []
        empleados.append(empleado)
        with open(self.nombre_archivo, 'wb') as f:
            pickle.dump(empleados, f)
        
        print(f"Empleado guardado: {empleado.nombre}")
    
    def busca_empleado(self, nombre):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                empleados = pickle.load(f)
                for emp in empleados:
                    if emp.nombre.lower() == nombre.lower():
                        return emp
        except FileNotFoundError:
            print("Archivo no encontrado")
        except EOFError:
            print("Archivo vacío")
        return None
    
    def mayor_salario(self, sueldo):
        try:
            with open(self.nombre_archivo, 'rb') as f:
                empleados = pickle.load(f)
                for emp in empleados:
                    if emp.salario > sueldo:
                        return emp
        except FileNotFoundError:
            print("Archivo no encontrado")
        except EOFError:
            print("Archivo vacío")
        return None

if __name__ == "__main__":
    archivo = ArchivoEmpleado("empleados.dat")
    archivo.guardar_empleado(Empleado("Juan Pérez", 2500.50))
    archivo.guardar_empleado(Empleado("María Gómez", 3200.75))
    archivo.guardar_empleado(Empleado("Carlos Ruiz", 1800.25))
    encontrado = archivo.busca_empleado("maría gómez")
    print(f"Empleado encontrado: {encontrado}")
    mayor_salario = archivo.mayor_salario(2000.0)
    print(f"Empleado con salario > 2000: {mayor_salario}")